File: easy/1160_find_words_that_can_be_formed_by_characters/solution.py
from collections import Counter
from typing import List


class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        chars_counter = Counter(chars)
        chars_count = 0

        for word in words:
            word_counter = Counter(word)

            for c, count in word_counter.items():
                if c not in chars_counter or chars_counter[c] < count:
                    break
            else:
                chars_count += len(word)

        return chars_count

File: easy/1160_find_words_that_can_be_formed_by_characters/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  int countCharacters(std::vector<std::string> &words, std::string chars) {
    std::vector<int> charsCounter(26, 0), charsCounterCopy;

    for (const char &ch : chars)
      ++charsCounter[ch - 'a'];

    bool isGood;
    int charsCount = 0;

    for (const std::string &word : words) {
      charsCounterCopy = charsCounter;
      isGood = true;

      for (const char &ch : word) {
        if (charsCounterCopy[ch - 'a'] == 0) {
          isGood = false;
          break;
        }

        --charsCounterCopy[ch - 'a'];
      }

      if (isGood)
        charsCount += word.size();
    }

    return charsCount;
  }
};
